libirods_rule_engine_plugin-audit_amqp.cpp

```diff
@@ -289,86 +322,95 @@ irods::error exec_rule(
-    gethostname( host_name, MAX_NAME_LEN );
-    json_object_set(
-        obj,
-        "hostname",
-        json_string(host_name));
+            // serialize the parameter to a map
+            irods::re_serialization::serialized_parameter_t param;
+            irods::error ret = irods::re_serialization::serialize_parameter(itr, param);
+            if(!ret.ok()) {
+                 rodsLog(
+                     LOG_ERROR,
+                     "unsupported argument for calling re rules from the rule language");
+                 continue;
+            }
 
-    pid_t pid = getpid();
-    std::stringstream pid_str; pid_str << pid;
-    json_object_set(
-        obj,
-        "pid",
-        json_string(pid_str.str().c_str()));
-
-    json_object_set(
-        obj,
-        "rule_name",
-        json_string(_rn.c_str()));
-
-    for( auto itr : _ps ) {
-        // The BytesBuf parameter should not be serialized because this commonly contains
-        // the entirety of the contents of files. These could be very big and cause the
-        // message broker to explode.
-        if (std::type_index(typeid(BytesBuf*)) == std::type_index(itr.type())) {
-            rodsLog(LOG_DEBUG9, "[{}:{}] - skipping serialization of BytesBuf parameter",
-                __FILE__, __LINE__);
-            continue;
-        }
+            for( auto elem : param ) {
 
-        // serialize the parameter to a map
-        irods::re_serialization::serialized_parameter_t param;
-        irods::error ret = irods::re_serialization::serialize_parameter(itr, param);
-        if(!ret.ok()) {
-             rodsLog(
-                 LOG_ERROR,
-                 "unsupported argument for calling re rules from the rule language");
-             continue;
-        }
+                size_t ctr = insert_arg_into_counter_map(arg_type_map, elem.first);
+                std::stringstream ctr_str;
+                ctr_str << ctr;
 
-        for( auto elem : param ) {
+                std::string key = elem.first;
+                if (ctr > 1) {
+                    key += "__";
+                    key += ctr_str.str();
+                }
 
-            size_t ctr = insert_arg_into_counter_map(arg_type_map, elem.first);
-            std::stringstream ctr_str;
-            ctr_str << ctr;
-            
-            std::string key = elem.first;
-            if (ctr > 1) {
-                key += "__";
-                key += ctr_str.str();
+            using byte_container_with_subtype = nlohmann::byte_container_with_subtype<std::vector<std::uint8_t>>;
+            try {
+                json_obj[key] = nlohmann::json::parse("\"" + elem.second + "\""); // force nlohmann to throw here if bad utf
+            } catch(const nlohmann::json::exception& e) {
+                json_obj[key] = byte_container_with_subtype(
+                    std::vector<std::uint8_t>(elem.second.begin(), elem.second.end()) // anything we don't understand is binary
+                );
```

Comment:
Is this different from using `nlohmann::json::binary(...)`?

Reply:
Oops, I missed this notification. It seems there is a difference, although in this case the end product would be the same. nlohmann::json::binary creates a "binary array value from a given binary container" whereas nlohmann:json::byte_container_with_subtype creates that binary container. Will change to the binary construct if for no other reason than clarity.